python/lsst/ip/isr/crosstalk.py

```diff
@@ -23,17 +23,441 @@
+
+    def toTable(self):
+        """Construct a list of tables containing the information in this calibration.
+
+        The list of tables should create an identical calibration
+        after being passed to this class's fromTable method.
+
+        Returns
+        -------
+        tableList : `list` [`lsst.afw.table.Table`]
+            List of tables containing the crosstalk calibration
+            information.
+
+        """
+        tableList = []
+        self.updateMetadata()
+        catalog = Table([{'CT_COEFFS': self.coeffs.reshape(self.nAmp*self.nAmp)}])
+        catalog.meta = self.getMetadata().toDict()
+        tableList.append(catalog)
+
+        if self.interChip:
+            interChipTable = Table([{'IC_SOURCE_DET': sourceDet,
+                                     'IC_COEFFS': self.interChip[sourceDet].reshape(self.nAmp*self.nAmp)}
+                                    for sourceDet in self.interChip.keys()])
+            tableList.append(interChipTable)
+        return tableList
+
+    # Implementation methods.
+    def extractAmp(self, image, amp, ampTarget, isTrimmed=False):
+
+        """Extract the image data from an amp, flipped to match ampTarget.
+
+        Parameters
+        ----------
+        image : `lsst.afw.image.Image` or `lsst.afw.image.MaskedImage`
+            Image containing the amplifier of interest.
+        amp : `lsst.afw.cameraGeom.Amplifier`
+            Amplifier on image to extract.
+        ampTarget : `lsst.afw.cameraGeom.Amplifier`
+            Target amplifier that the extracted image will be flipped
+            to match.
+        isTrimmed : `bool`
+            The image is already trimmed.
+            This should no longer be needed once DM-15409 is resolved.
+
+        Returns
+        -------
+        output : `lsst.afw.image.Image`
+            Image of the amplifier in the desired configuration.
+        """
+        X_FLIP = {lsst.afw.cameraGeom.ReadoutCorner.LL: False,
+                  lsst.afw.cameraGeom.ReadoutCorner.LR: True,
+                  lsst.afw.cameraGeom.ReadoutCorner.UL: False,
+                  lsst.afw.cameraGeom.ReadoutCorner.UR: True}
+        Y_FLIP = {lsst.afw.cameraGeom.ReadoutCorner.LL: False,
+                  lsst.afw.cameraGeom.ReadoutCorner.LR: False,
+                  lsst.afw.cameraGeom.ReadoutCorner.UL: True,
+                  lsst.afw.cameraGeom.ReadoutCorner.UR: True}
+
+        output = image[amp.getBBox() if isTrimmed else amp.getRawDataBBox()]
+        thisAmpCorner = amp.getReadoutCorner()
+        targetAmpCorner = ampTarget.getReadoutCorner()
+
+        # Flipping is necessary only if the desired configuration doesn't match what we currently have
+        xFlip = X_FLIP[targetAmpCorner] ^ X_FLIP[thisAmpCorner]
+        yFlip = Y_FLIP[targetAmpCorner] ^ Y_FLIP[thisAmpCorner]
+        self.log.debug("Extract amp: %s %s %s %s",
+                       amp.getName(), ampTarget.getName(), thisAmpCorner, targetAmpCorner)
```

Comment:
f-strings preferred

Reply:
I thought they were not for lsst.Log messages, as the log only applies the formatting if the log level triggers?

Reply:
My two bits, as an kibbitzer: I agree that deferred formatting can be helpful in log messages, especially if formatting is expensive and the messages are unlikely to be output (e.g. debug level, as here).

That said, I agree with @mrawls that f strings are better for log messages that are simple or usually output (e.g. info and warning level), because the code is a always exercised and you don't have to wonder if the formatting is correct.

Reply:
I'm going to leave this one alone (partially to avoid having to deal with a long line), and use f-strings for info and warning going forward.
I may try to bother someone about getting a definite answer put into the style guide, so I can either get it right or be consistently wrong until I get it right.

Reply:
Didn't mean to open a pandora's box about string formatting here 😄 not a big deal in this particular instance, but agree it should be in the style guide one way or another.